seq2science/rules/quantification.smk

```diff
@@ -166,22 +166,45 @@ elif config["quantifier"] == "kallistobus":
             {params.options} > {log} 2>&1
             """
 
-    def get_kallistobus_reads(wildcards):
-        reads = []
-        sample = wildcards.sample
+    def get_fastq_pair_reads(wildcards):
+        reads = dict()
         assert sampledict[sample]["layout"] == "PAIRED"
         read_id = get_bustools_rid(config.get("count"))
         #Determine mate for trimming
         if read_id == 0:
-            reads += expand(f"{{trimmed_dir}}/{sample}_R1_trimmed.{{fqsuffix}}.gz", **config)
-            reads += expand("{fastq_dir}/{{sample}}_R2.{fqsuffix}.gz", **config)
+            reads["r1"] = expand("{trimmed_dir}/{{sample}}_{fqext1}_trimmed.{fqsuffix}.gz", **config)
+            reads["r2"] = expand("{fastq_dir}/{{sample}}_{fqext2}.{fqsuffix}.gz", **config)
         elif read_id == 1:
-            reads += expand("{fastq_dir}/{{sample}}_R1.{fqsuffix}.gz", **config)
-            reads += expand(f"{{trimmed_dir}}/{sample}_R2_trimmed.{{fqsuffix}}.gz", **config)
-        else:    
+            reads["r1"] = expand("{fastq_dir}/{{sample}}_{fqext1}.{fqsuffix}.gz", **config)
+            reads["r2"] = expand("{trimmed_dir}/{{sample}}_{fqext2}_trimmed.{fqsuffix}.gz", **config)
+        else:
             raise NotImplementedError
-
         return reads
+        
+    rule fastq_pair:    
+        """
+        Example
+        """
+        input:
+            unpack(get_fastq_pair_reads)
+        output:
+            r1=expand("{fastq_clean_dir}/{{sample}}_clean_{fqext1}.{fqsuffix}.paired.fq", **config),
+            r2=expand("{fastq_clean_dir}/{{sample}}_clean_{fqext2}.{fqsuffix}.paired.fq", **config),
+            intermediates1=temp(expand("{fastq_clean_dir}/{{sample}}_clean_{fqext}.{fqsuffix}", **config)),             
+            intermediates2=temp(expand("{fastq_clean_dir}/{{sample}}_clean_{fqext}.{fqsuffix}{singles}.fq", **{**config,
+                                                                                                            **{"singles": [".single"]}}))                                                                               
+        priority: 1
+        conda:
+            "../envs/fastq-pair.yaml"
+        params:
+            clean_dir=config.get("fastq_clean_dir"),
+            options=config.get("fastq-pair")
+        shell:
+            """
+            gunzip -c {input.r1} > {params.clean_dir}/{wildcards.sample}_clean_R1.fastq 
+            gunzip -c {input.r2} > {params.clean_dir}/{wildcards.sample}_clean_R2.fastq
+            fastq_pair {params.options} {params.clean_dir}/{wildcards.sample}_clean_R1.fastq {params.clean_dir}/{wildcards.sample}_clean_R2.fastq
```

Comment:
If I got it correctly, you use `gunzip`  to make the intermediate files? If so, you could replace it with 
```
gunzip -c {input.r1} > {output.intermediates1}
gunzip -c {input.r2} > {output.intermediates2}
fastq_pair {params.options} {output.intermediates1} {output.intermediates2}
```
saves you a param as well!

Reply:
Maybe the -t can be derived dynamically, something like this?

```
wc -l {input.r1} | grep -Po "^\d+" | awk '{print int($1/4)}'
```

If you want to be really fancy you could make it so that if the user specifies a `-t` you use theirs, if they don't you use the dynamicly derived one.

Untested:

```
options=lambda wildcards, input: config.get("fastq-pair", "") if "-t" in config.get("fastq-pair", "") else config.get("fastq-pair", "") + " " + f"-t \"$(wc -l {input.r1} | grep -Po "^\d+" | awk '{{print int($1/4)}}'\""
```

Needs some cleaning up :yum:

Reply:
Good one. Now it determines -t dynamically or derives them from the config settings (if present)